### PR TITLE
Fix incorrect conversion on overflow.

### PR DESCRIPTION
**Pull request by racardoso**
This commit fixes incorrect conversions between decimal and binary types that
always produce an infinity for obviously overflowing results, without taking
account of the rounding mode (task #31).

Signed-off-by: Rogerio Alves <rcardoso@linux.vnet.ibm.com>

### CODE REVIEW

**Comment by racardoso** [comment]
I am working on the underflow problem too. But I am willing to hear any reviews on this before. This is my first PR on this project.

File: base-math/truncdddf.c [hunk 1]
```diff
@@ -49,7 +52,20 @@ CONVERT_WRAPPER(
 	  {
 	    if (DFP_EXCEPTIONS_ENABLED)
 	      DFP_HANDLE_EXCEPTIONS (FE_OVERFLOW|FE_INEXACT);
-	    return SIGNBIT(a) ? -INFINITY : INFINITY;
+
+		/* Default result when trap is disable - IEEE754 */
```
Comment (gftg85) [inline base-math/truncdddf.c]: I did not understand this comment...  Did you mean something in the lines of:

Correctly round the result according to the current rounding direction mode.

File: base-math/truncdddf.c [hunk 2]
```diff
@@ -49,7 +52,20 @@ CONVERT_WRAPPER(
...
+	      {
+	        case FE_TOWARDZERO:
+	          return SIGNBIT(a) ? -DBL_MAX : DBL_MAX;
+	        case FE_DOWNWARD:
+	          return SIGNBIT(a) ? -INFINITY : DBL_MAX;
+	        case FE_UPWARD:
+	          return SIGNBIT(a) ? -DBL_MAX : INFINITY;
+	        case FE_TONEAREST:
+	        default:
+	          return SIGNBIT(a) ? -INFINITY : INFINITY;
```
Comment (gftg85) [inline base-math/truncdddf.c]: In lines 59 to 67, replace 8 spaces with 1 tab.

File: base-math/truncddsf.c [hunk 3]
```diff
@@ -38,16 +41,29 @@ CONVERT_WRAPPER(
 	_Decimal64 a_norm;
 	long long mant;
 	int exp, sexp;
-	
+
```
Comment (gftg85) [inline base-math/truncddsf.c]: I'd prefer that this change, which is not related to the rounding direction mode problem, not be integrated.

File: base-math/truncddsf.c [hunk 4]
```diff
@@ -38,16 +41,29 @@ CONVERT_WRAPPER(
...
 	a_norm = getmantandexpd64 (a, &exp, 16, 1e16DD);
-	
+
```
Comment (gftg85) [inline base-math/truncddsf.c]: Likewise.

File: base-math/truncddsf.c [hunk 5]
```diff
@@ -38,16 +41,29 @@ CONVERT_WRAPPER(
...
+	      {
+	        case FE_TOWARDZERO:
+	          return SIGNBIT(a) ? -FLT_MAX : FLT_MAX;
+	        case FE_DOWNWARD:
+	          return SIGNBIT(a) ? -INFINITY : FLT_MAX;
+	        case FE_UPWARD:
+	          return SIGNBIT(a) ? -FLT_MAX : INFINITY;
+	        case FE_TONEAREST:
+	        default:
+	          return SIGNBIT(a) ? -INFINITY : INFINITY;
```
Comment (gftg85) [inline base-math/truncddsf.c]: In lines 57 to 65, replace 8 spaces with 1 tab.

File: base-math/trunctddf.c [hunk 6]
```diff
@@ -48,7 +51,20 @@ CONVERT_WRAPPER(
...
+	      {
+	        case FE_TOWARDZERO:
+	          return SIGNBIT(a) ? -DBL_MAX : DBL_MAX;
+	        case FE_DOWNWARD:
+	          return SIGNBIT(a) ? -INFINITY : DBL_MAX;
+	        case FE_UPWARD:
+	          return SIGNBIT(a) ? -DBL_MAX : INFINITY;
+	        case FE_TONEAREST:
+	        default:
+	          return SIGNBIT(a) ? -INFINITY : INFINITY;
```
Comment (gftg85) [inline base-math/trunctddf.c]: In lines 58 to 66, replace 8 spaces with 1 tab.

File: base-math/trunctdsf.c [hunk 7]
```diff
@@ -38,15 +41,28 @@ CONVERT_WRAPPER(
 	SRC_TYPE a_norm;
 	long long mant;
 	int	exp, sexp;
-	
+
```
Comment (gftg85) [inline base-math/trunctdsf.c]: Same comment about unrelated changes.

File: base-math/trunctdsf.c [hunk 8]
```diff
@@ -38,15 +41,28 @@ CONVERT_WRAPPER(
...
+	      {
+	        case FE_TOWARDZERO:
+	          return SIGNBIT(a) ? -FLT_MAX : FLT_MAX;
+	        case FE_DOWNWARD:
+	          return SIGNBIT(a) ? -INFINITY :FLT_MAX;
+	        case FE_UPWARD:
+	          return SIGNBIT(a) ? -FLT_MAX : INFINITY;
+	        case FE_TONEAREST:
+	        default:
+	          return SIGNBIT(a) ? -INFINITY : INFINITY;
```
Comment (gftg85) [inline base-math/trunctdsf.c]: In lines 56 to 64, replace 8 spaces with 1 tab.

File: base-math/truncdddf.c (hunk 1 above)
Comment (racardoso) [reply base-math/truncdddf.c]: @gftg85  That weird comment I forgot to remove. That's was for remind me that POWER ISA 2.07 says (pag. 125) we should adjust the exponent of the intermediate result and adjust the round result if the overflow exception is enable. If it's disable we should do as I implement. It's not clear to me if in this case "obvious overflow" when cast to double if we need to adjust the exponent and return as "Normal Number" or just do what I am doing.

File: base-math/truncdddf.c (hunk 2 above)
Comment (racardoso) [reply base-math/truncdddf.c]: @gftg85 I was following how it's implemented on the rest of the file. On my vim it appears (set list) as one tab and spaces. But ok I'll check that.

**Comment by racardoso** [comment]
Tasks: 

- [ ] Add a test to avoid regressions.

File: base-math/truncdddf.c (hunk 2 above)
Comment (gftg85) [reply base-math/truncdddf.c]: Hrm, you are right.  Please disregard this comment.

It seems that there isn't a coding style being followed and the relevant section in README.developer is in TODO state.

Sorry about the inconvenience.

File: base-math/truncdddf.c (hunk 1 above)
Comment (gftg85) [reply base-math/truncdddf.c]: Section 4.4.3 of POWER ISA 2.07 only refers to operations with binary floating-point formats.  Section 5.5.10.4 refers to operations with decimal floating-point formats.  None of them describe how conversions between these floating-point types work. Moreover, they are specific to POWER.

On the other hand, ISO-IEC TR 24732, in section 6.2, describes how a C implementation of decimal floating-point type should make the conversions between floating-point types.  Based on that, I think that your implementation is correct.

File: base-math/truncddsf.c (hunk 3 above)
Comment (racardoso) [reply base-math/truncddsf.c]: I agree. But notice that for base-math/trunctddf.c there's no tab on those lines. Maybe I should do another PR to fix that.

File: base-math/truncddsf.c (hunk 4 above)
Comment (racardoso) [reply base-math/truncddsf.c]: Likewise.

File: base-math/trunctdsf.c (hunk 7 above)
Comment (racardoso) [reply base-math/trunctdsf.c]: Same thing about differences between trunc*** files. Is this tab incorrect?

**Comment by gftg85** [comment]
Your white-space fixes seem reasonable.  My only concern is that they have nothing to do with the overflow problem, so I'd like to see these changes removed from this patch.

If you want to fix the white-space changes in another commit, that's fine by me.

File: base-math/truncddsf.c [hunk 9]
```diff
@@ -40,14 +43,26 @@ CONVERT_WRAPPER(
 	int exp, sexp;
 	
 	a_norm = getmantandexpd64 (a, &exp, 16, 1e16DD);
-	
+
```
Comment (gftg85) [inline base-math/truncddsf.c]: Thanks for removing the other occurrences of this change.
This one slipped in...  Can you remove it, too?

File: tests/test-cast-to-overflow.c [hunk 10]
```diff
@@ -0,0 +1,91 @@
+
```
Comment (gftg85) [inline tests/test-cast-to-overflow.c]: This empty line is not necessary and I don't see it in other files.
Could you remove it?

File: tests/test-cast-to-overflow.c [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+   General Public License version 2.1.
+   The Decimal Floating Point C Library is distributed in the hope that
+   it will be useful, but WITHOUT ANY WARRANTY; without even the implied
+   warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
+   the GNU Lesser General Public License version 2.1 for more details.
+   You should have received a copy of the GNU Lesser General Public
+   License version 2.1 along with the Decimal Floating Point C Library;
+   if not, write to the Free Software Foundation, Inc., 59 Temple Place,
+   Suite 330, Boston, MA 02111-1307 USA.
+   Please see libdfp/COPYING.txt for more information.  */
```
Comment (gftg85) [inline tests/test-cast-to-overflow.c]: This license block has a weird format.
Could you re-write it so that it looks similar to the other files?

File: tests/test-cast-to-overflow.c [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+  for (i=0; i<d128to32_size; ++i)
+    {
+      float ret;
+      fesetround (d128to32[i].round_mode);
+      ret = (float) d128to32[i].set;
+      _VC_P (__FILE__, d128to32[i].line, d128to32[i].expect, ret, "%a");
+    }
+
```
Comment (gftg85) [inline tests/test-cast-to-overflow.c]: Do we want to test for long double, too?

File: base-math/truncddsf.c (hunk 9 above)
Comment (racardoso) [reply base-math/truncddsf.c]: Oops. I will remove it.

File: tests/test-cast-to-overflow.c (hunk 10 above)
Comment (racardoso) [reply tests/test-cast-to-overflow.c]: Yes.

File: tests/test-cast-to-overflow.c (hunk 11 above)
Comment (racardoso) [reply tests/test-cast-to-overflow.c]: I did a ctrl+c ctrl+v on the license from another file? What's wrong? Indentation?

File: tests/test-cast-to-overflow.c (hunk 12 above)
Comment (racardoso) [reply tests/test-cast-to-overflow.c]: Isn't long double and double the same on POWER?

File: tests/test-cast-to-overflow.c (hunk 11 above)
Comment (gftg85) [reply tests/test-cast-to-overflow.c]: At a glance it seems different.

I see that there are empty lines splitting the paragraphs in the other files.  Maybe that's the only difference.

PS: I just noticed that the year is 2016.  Please change it to 2017.

File: tests/test-cast-to-overflow.c (hunk 12 above)
Comment (gftg85) [reply tests/test-cast-to-overflow.c]: No.  It might have been in the past, but it is not anymore.  On POWER, long double uses two 64-bits registers to get more precision (without changing the range, i.e.: no changes in the exponent size).

https://en.wikipedia.org/wiki/IBM_Floating_Point_Architecture#Extended-precision_128-bit

Comment (racardoso) [reply tests/test-cast-to-overflow.c]: Turns out we found a new bug. Conversion to long double returns a NaN instead Infinity. I've opened a new task (#60) to fix this problem. It's a different kind of problem since on that case it's not related to the rounding mode but to the conversion itself. (It's something before the rounding mode switch).

Comment (gftg85) [reply tests/test-cast-to-overflow.c]: OK.  Thanks for the new bug report.
You don't need to address that in this pull request.

**Comment by gftg85** [comment]
Thanks for the new patch with the fixes...

Before I can approve this, could you rebase it and turn these three commits into a single commit or into two commits (one for the fix and the other for the test case)?
